File: programmers/level2/게임 맵 최단거리.py
import collections

def solution(maps):
    answer = 0
    queue = collections.deque()
    queue.append((0,0,1))
    setpos = set()
    setpos.add((0,0))
    movex = [1,0,0,-1]
    movey = [0,1,-1,0]
    while queue:
        pos_y, pos_x, now_cnt = queue.popleft()
        for i in range(4):
            next_y, next_x = pos_y+movey[i], pos_x+movex[i]
            if next_x==len(maps[0])-1 and next_y==len(maps)-1: return now_cnt+1
            if 0<=next_x<len(maps[0]) and 0<=next_y<len(maps) and maps[next_y][next_x]!=0 and (next_y,next_x) not in setpos:
                queue.append((next_y,next_x, now_cnt+1))
                setpos.add((next_y, next_x))
    
    return -1

File: programmers/level2/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        Queue<Point> q = new LinkedList<>();
        q.offer(new Point(0, 0, 1));
        int[] yMove = {1,-1,0,0};
        int[] xMove = {0,0,1,-1};
        int[][] visited = new int[maps.length][maps[0].length];
        visited[0][0] = 1;
        
        while(q.size() > 0) {
            Point myPos = q.poll();
            if(myPos.y == maps.length-1 && myPos.x == maps[0].length-1) return myPos.cnt;
            
            for(int i=0; i<4; i++) {
                int nextY = myPos.y + yMove[i];
                int nextX = myPos.x + xMove[i];
                int nextCnt = myPos.cnt + 1;
                boolean check = (0<=nextY && nextY<maps.length && 0<=nextX && nextX<maps[0].length && visited[nextY][nextX] == 0 && maps[nextY][nextX] == 1);
                if(check) {
                    Point nextPoint = new Point(nextY, nextX, nextCnt);
                    q.offer(nextPoint);
                    visited[nextY][nextX] = 1;
                }
            }
        }
        
        return -1;
    }
    
    class Point {
        int y;
        int x;
        int cnt;
        
        Point(int y, int x, int cnt) {
            this.y = y;
            this.x = x;
            this.cnt = cnt;
        }
    };
}